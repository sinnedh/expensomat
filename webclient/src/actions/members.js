export const getMembers = (token) => {
  return {type: 'MEMBERS:LOAD_REQUEST', token}
}

export const createMember = (token, member) => {
  return {type: 'MEMBERS:CREATE_REQUEST', token, member}
}

export const deleteMember = (token, member) => {
  return {type: 'MEMBERS:DELETE_REQUEST', token, member}
}

export const updateMember = (token, id, changes) => {
  return {type: 'MEMBERS:UPDATE_REQUEST', token, id, changes}
}
