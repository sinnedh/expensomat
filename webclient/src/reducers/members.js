import { List, Map } from 'immutable'

const initialState = Map()

const members = (state = initialState, action) => {
  switch (action.type) {
    case 'MEMBERS:LOAD_SUCCESS':
      return List(action.items).reduce(
        (result, value) => result.set(value['id'].toString(), Map(value)),
        Map()
      )

    case 'MEMBERS:CREATE_SUCCESS':
      return state.merge({[action.member.id]: Map({...action.member})})

    case 'MEMBERS:DELETE_SUCCESS':
      return state.delete(action.id.toString())

    case 'MEMBERS:UPDATE_SUCCESS':
      return state.merge({[action.id]: {name: action.name}})

    case 'MEMBERS:LOAD_REQUEST':
    case 'MEMBERS:LOAD_FAILURE':
    case 'MEMBERS:CREATE_REQUEST':
    case 'MEMBERS:CREATE_FAILURE':
    case 'MEMBERS:DELETE_REQUEST':
    case 'MEMBERS:DELETE_FAILURE':
    case 'MEMBERS:UPDATE_REQUEST':
    case 'MEMBERS:UPDATE_FAILURE':
    default:
      return state;
  }
}

export default members;
