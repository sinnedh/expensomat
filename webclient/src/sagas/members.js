import { call, put } from 'redux-saga/effects'
import { setErrorNotification, setInfoNotification } from '../actions'
import { incrementLoadingCounter, decrementLoadingCounter } from '../actions'
import * as api from '../api'

export function* createMember(action) {
  yield put(incrementLoadingCounter())
  try {
    const member = yield call(api.createMember, action.token, action.member)
    yield put({
      type: "MEMBERS:CREATE_SUCCESS",
      member: member.data,
    })
    yield put(setInfoNotification('Member created succesfully'))
  } catch (e) {
    yield put({type: "MEMBERS:CREATE_FAILURE", message: e.message})
    yield put(setErrorNotification(`Could not create member ("${e.message}")`))
  } finally {
    yield put(decrementLoadingCounter())
  }
}

export function* updateMember(action) {
  yield put(incrementLoadingCounter())
  try {
    const member = yield call(api.updateMember, action.token, action.id, action.changes)
    yield put({
      type: "MEMBERS:UPDATE_SUCCESS",
      id: member.data.id,
      name: member.data.name,
    })
    yield put(setInfoNotification('Member updated succesfully'))
  } catch (e) {
    yield put({type: "MEMBERS:UPDATE_FAILURE", message: e.message})
    yield put(setErrorNotification(`Could not update member ("${e.message}")`))
  } finally {
    yield put(decrementLoadingCounter())
  }
}

export function* deleteMember(action) {
  yield put(incrementLoadingCounter())
  try {
    yield call(api.deleteMember, action.token, action.memberId)
    yield put({
      type: "MEMBERS:DELETE_SUCCESS",
      id: action.memberId,
    })
    yield put(setInfoNotification('Member deleted succesfully'))
  } catch (e) {
    yield put({type: "MEMBERS:DELETED_FAILURE", message: e.message})
    yield put(setErrorNotification(`Could not delete member ("${e.message}")`))
  } finally {
    yield put(decrementLoadingCounter())
  }
}

export function* fetchMembers(action) {
  yield put(incrementLoadingCounter())

  try {
    const members = yield call(api.fetchMembers, action.token)
    yield put({
      type: "MEMBERS:LOAD_SUCCESS",
      items: members.data,
    })
  } catch (e) {
    yield put({type: "MEMBERS:LOAD_FAILURE", message: e.message})
    yield put(setErrorNotification(`Could not load members ("${e.message}")`))
  } finally {
    yield put(decrementLoadingCounter())
  }
}
