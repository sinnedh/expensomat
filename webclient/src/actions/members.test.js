import * as actions from './members'

describe('MEMBERS', () => {
  const token = 'ABCD123'
  const member = {id: 1, name: 'Member'}

  it('handles getExpenses', () => {
    const action = actions.getMembers(token)
    expect(action).toMatchSnapshot()
  })

  it('handles createMember', () => {
    const action = actions.createMember(token, member)
    expect(action).toMatchSnapshot()
  })

  it('handles deleteMember', () => {
    const action = actions.deleteMember(token, member)
    expect(action).toMatchSnapshot()
  })

  it('handles updateMember', () => {
    const action = actions.updateMember(token, 11, {name: 'Member'})
    expect(action).toMatchSnapshot()
  })
})
