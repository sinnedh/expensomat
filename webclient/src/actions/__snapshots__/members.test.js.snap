// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MEMBERS handles createMember 1`] = `
Object {
  "member": Object {
    "id": 1,
    "name": "Member",
  },
  "token": "ABCD123",
  "type": "MEMBERS:CREATE_REQUEST",
}
`;

exports[`MEMBERS handles deleteMember 1`] = `
Object {
  "member": Object {
    "id": 1,
    "name": "Member",
  },
  "token": "ABCD123",
  "type": "MEMBERS:DELETE_REQUEST",
}
`;

exports[`MEMBERS handles getExpenses 1`] = `
Object {
  "token": "ABCD123",
  "type": "MEMBERS:LOAD_REQUEST",
}
`;

exports[`MEMBERS handles updateMember 1`] = `
Object {
  "changes": Object {
    "name": "Member",
  },
  "id": 11,
  "token": "ABCD123",
  "type": "MEMBERS:UPDATE_REQUEST",
}
`;
